src/db.c

```diff
@@ -1065,34 +1086,37 @@ void scanGenericCommand(client *c, robj *o, unsigned long long cursor) {
      * just return everything inside the object in a single call, setting the
      * cursor to zero to signal the end of the iteration. */
 
-    /* Handle the case of a hash table. */
-    ht = NULL;
+    /* Handle the case of a dict or hashset. */
+    dict *dictTable = NULL;
+    hashset *hashsetTable = NULL;
     if (o == NULL) {
-        ht = NULL;
-    } else if (o->type == OBJ_SET && o->encoding == OBJ_ENCODING_HT) {
-        ht = o->ptr;
+        dictTable = NULL;
+    } else if (o->type == OBJ_SET && o->encoding == OBJ_ENCODING_HASHSET) {
+        hashsetTable = o->ptr;
     } else if (o->type == OBJ_HASH && o->encoding == OBJ_ENCODING_HT) {
-        ht = o->ptr;
+        dictTable = o->ptr;
     } else if (o->type == OBJ_ZSET && o->encoding == OBJ_ENCODING_SKIPLIST) {
         zset *zs = o->ptr;
-        ht = zs->dict;
+        dictTable = zs->dict;
     }
 
     list *keys = listCreate();
     /* Set a free callback for the contents of the collected keys list.
-     * For the main keyspace dict, and when we scan a key that's dict encoded
-     * (we have 'ht'), we don't need to define free method because the strings
-     * in the list are just a shallow copy from the pointer in the dictEntry.
+     * For the main keyspace dict, when we scan a key that's dict encoded
+     * (we have 'dictTable'), or when we scan a key that's hashset encoded
+     * (we have 'hashsetTable') we don't need to define free method because the
+     * strings in the list are just a shallow copy from the pointer in the
+     * dictEntry.
      * When scanning a key with other encodings (e.g. listpack), we need to
      * free the temporary strings we add to that list.
      * The exception to the above is ZSET, where we do allocate temporary
      * strings even when scanning a dict. */
-    if (o && (!ht || o->type == OBJ_ZSET)) {
+    if (o && ((!dictTable && !hashsetTable) || o->type == OBJ_ZSET)) {
```

Comment:
I wonder if this is the clearest way to express this. It seems to me that just checking `o->encoding` directly would be more clear.

```suggestion
    if (o && ((o->encoding != OBJ_ENCODING_HT && o->encoding != OBJ_ENCODING_HASHSET) || o->type == OBJ_ZSET)) {
```

Reply:
You're right, this could be clearer. I think setting a list_items_shallow_copied int in the if/else above will be most clear.

fwiw, once all the data types are ported to hashset the parallel dict/hashset stuff like this will be gone

Reply:
Yeah good point.